stlab/concurrency/wait.hpp

```diff
@@ -0,0 +1,223 @@
+/*
+    Copyright 2015 Adobe
+    Distributed under the Boost Software License, Version 1.0.
+    (See accompanying file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+*/
+
+/**************************************************************************************************/
+
+#ifndef STLAB_CONCURRENCY_WAIT_HPP
+#define STLAB_CONCURRENCY_WAIT_HPP
+
+#include <chrono>
+#include <condition_variable>
+#include <exception>
+#include <mutex>
+#include <type_traits>
+#include <utility>
+
+#include <stlab/concurrency/default_executor.hpp>
+#include <stlab/concurrency/future.hpp>
+#include <stlab/concurrency/immediate_executor.hpp>
+#include <stlab/concurrency/ready_future.hpp>
+#include <stlab/memory.hpp>
+
+/**************************************************************************************************/
+
+namespace stlab {
+
+/**************************************************************************************************/
+
+inline namespace v1 {
+
+/**************************************************************************************************/
+
+/**
+    Assumes f _will wait_ and wakes or adds a thread to the thread pool (to the limit) before
+    invoking f.
+*/
+template <class F>
+auto invoke_waiting(F&& f) {
+#if STLAB_TASK_SYSTEM(PORTABLE)
+    if (!detail::pts().wake()) detail::pts().add_thread();
+#endif
+
+    std::forward<F>(f)();
+}
+
+/**************************************************************************************************/
+
+namespace detail {
+
+template <class T>
+struct _get_ready_future {
+    template <class F>
+    auto operator()(F&& f) {
+        return *std::forward<F>(f).get_try();
+    }
+};
+
+template <>
+struct _get_ready_future<void> {
+    template <class F>
+    void operator()(F&& f) {
+        std::forward<F>(f).get_try();
+    }
+};
+
+template <class T>
+struct _get_optional;
+
+template <class T>
+struct _get_optional<stlab::optional<T>> {
+    template <class F>
+    auto operator()(F&& f) {
+        return *std::forward<F>(f);
+    }
+};
+
+template <>
+struct _get_optional<bool> {
+    template <class F>
+    auto operator()(F&&) {}
+};
+
+} // namespace detail
+
+template <typename T>
+T get_wait(future<T> x) {
+    if (auto result = x.get_try())
+        return detail::_get_optional<decltype(result)>{}(std::move(result)); // if ready, done
+
+    std::mutex m;
+    std::condition_variable condition;
+    bool flag{false};
+
+    auto hold = std::move(x).recover(immediate_executor, [&](auto&& r) {
+        x = std::forward<decltype(r)>(r);
+        {
+            std::unique_lock<std::mutex> lock{m};
+            flag = true;
+            condition.notify_one(); // must notify under lock
+        }
+    });
+
+#if STLAB_TASK_SYSTEM(PORTABLE)
+
+    auto resolved = [&] {
+        std::unique_lock<std::mutex> lock{m};
+        return flag;
+    };
+
+    if (!resolved() && !detail::pts().wake() && !resolved()) detail::pts().add_thread();
```

Comment:
How big is the probability that the second !resolved() returns a different result than the first one? I assume that you want to delay as long as possible the creation of a new thread

Reply:
I agree - even the first is a very low probability. Simplified.